### PR TITLE
Polling support

### PR DESCRIPTION
**Pull request by artem-tverdokhlebov**
Polling support by @EmotionalLove

### CODE REVIEW

**Comment by artem-tverdokhlebov** [comment]
@kprinssu can you check this, please? :)

File: VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp [hunk 1]
```diff
@@ -29,6 +29,8 @@ bool VoodooI2CELANTouchpadDriver::check_ASUS_firmware(UInt8 productId, UInt8 ic_
 }
 
 bool VoodooI2CELANTouchpadDriver::init(OSDictionary *properties) {
+    OSSafeReleaseNULL(interrupt_simulator);
```
Comment (kprinssu) [inline VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp:32]: Why are you release `interrupt_simulator` on driver initialisation? What if `interrupt_simulator` is not NULL, then what would happen when release is called? It's undefined behaviour. Best to remove it here OR set to NULL.

Comment (kprinssu) [reply VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp:32]: Additionally, you should call `super::init`.

File: VoodooI2CELAN/VoodooI2CELANTouchpadDriver.hpp [hunk 2]
```diff
@@ -176,6 +180,8 @@ class VoodooI2CELANTouchpadDriver : public IOService {
      * @return kIOSuccess if the message is processed
      */
     IOReturn message(UInt32 type, IOService* provider, void* argument) override;
+    
+    void simulateInterrupt(OSObject* owner, IOTimerEventSource* timer); // Sasha - impl polling support
```
Comment (kprinssu) [inline VoodooI2CELAN/VoodooI2CELANTouchpadDriver.hpp:184]: Please remove the  @EmotionalLove's comment here.

File: VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp (hunk 1 above)
Comment (artem-tverdokhlebov) [reply VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp]: fixed

File: VoodooI2CELAN/VoodooI2CELANTouchpadDriver.hpp (hunk 2 above)
Comment (artem-tverdokhlebov) [reply VoodooI2CELAN/VoodooI2CELANTouchpadDriver.hpp]: fixed

**Comment by kprinssu** [comment]
@artem-tverdokhlebov this looks good to me, I'll let @ben9923 take a second look.

File: VoodooI2CELAN/VoodooI2CELANTouchpadDriver.hpp [hunk 3]
```diff
@@ -8,8 +8,10 @@
 
 #ifndef VOODOOI2C_ELAN_TOUCHPAD_DRIVER_HPP
 #define VOODOOI2C_ELAN_TOUCHPAD_DRIVER_HPP
+#define INTERRUPT_SIMULATOR_TIMEOUT 5
```
Comment (ben9923) [inline VoodooI2CELAN/VoodooI2CELANTouchpadDriver.hpp:11]: Don't put any definition next to the include guard. Put it after the `ELAN_NAME` definition instead, for example :)

File: VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp [hunk 4]
```diff
@@ -491,17 +498,32 @@ bool VoodooI2CELANTouchpadDriver::start(IOService* provider) {
...
+        IOLog("%s::%s Could not get interrupt event source, trying to fallback on polling\n", getName(), elan_name);
+        interrupt_simulator = IOTimerEventSource::timerEventSource(this, OSMemberFunctionCast(IOTimerEventSource::Action, this, &VoodooI2CELANTouchpadDriver::simulateInterrupt));
+        if (!interrupt_simulator) {
+            IOLog("%s::%s Could not get timer event source\n", getName(), elan_name);
+            goto start_exit;
+        }
+        publish_multitouch_interface();
+        if (!init_device()) {
+            IOLog("%s::%s Failed to init device\n", getName(), elan_name);
+            return NULL;
```
Comment (ben9923) [inline VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp:512]: Don't you want to `goto start_exit` here? Same for L521.

File: VoodooI2CELAN/VoodooI2CELANTouchpadDriver.hpp (hunk 3 above)
Comment (artem-tverdokhlebov) [reply VoodooI2CELAN/VoodooI2CELANTouchpadDriver.hpp]: Done

File: VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp (hunk 4 above)
Comment (artem-tverdokhlebov) [reply VoodooI2CELAN/VoodooI2CELANTouchpadDriver.cpp]: Done

**Pull request merged**